source/agora/network/Client.d

```diff
@@ -102,6 +164,75 @@ public class NetworkClient
         this.exception = new Exception(
             format("Request failure to %s after %s attempts", address,
                 max_retries));
+        this.gossip_timer = this.taskman.setTimer(0.seconds, &this.gossipScheduler);
+    }
+
+    /// Shut down the gossiping timer
+    public void shutdown ()
+    {
+        this.gossip_timer.stop();
+    }
+
+    /// For gossiping we don't want to block the calling fiber, so we use
+    /// use request queueing and a separate fiber to handle all the outgoing requests.
+    private void gossipScheduler ()
+    {
+        while (1)
+        {
+            while (!this.gossip_queue.empty)
+            {
+                auto event = this.gossip_queue.front;
+                this.gossip_queue.removeFront();
+                this.handleGossip(event);
+                // yield w/ quick sleep return for next event
+                this.taskman.wait(0.msecs);
+            }
+
+            this.taskman.wait(10.msecs);  // no events, yield w/ longer sleep
+        }
+    }
+
+    /// Handle an outgoing gossip event
+    private void handleGossip (GossipEvent event) nothrow
+    {
+        switch (event.type) with (GossipType)
+        {
+        case Tx:
+            import agora.crypto.Hash;
+            const tx_hash = event.tx.hashFull();
+
+            // if the node already has this tx, don't send it
+            if (this.attemptRequest!(API.hasTransactionHash, Throw.No,
```

Comment:
I wonder if this is in fact slower as it is two API calls.

Reply:
It depends on latency, bandwidth, and the tx size. We are checking the hash first before we send the full tx. But if latency is large then yes this will block the fiber. Maybe we should just send without asking?

Reply:
Perhaps it can depend on the size of the `tx` which could be a configuration parameter. It doesn't have to be part of this PR though.

Reply:
We could perhaps take a look at what Bitcoin is doing.

Reply:
They broadcast unconditionally according to [docs here](https://github.com/bitcoin/bitcoin/blob/ed25cb58f605ba583c735f330482df0bf9348f3a/src/rpc/rawtransaction.cpp#L819)

Reply:
Yeah I was looking at the code and reached the same conclusion. I was the one who added this check here, so I'll just remove it and see what happens.

Reply:
🤞